influxdb3/src/commands/serve.rs

```diff
@@ -238,33 +244,40 @@ pub async fn command(config: Config) -> Result<()> {
         trace_exporter,
         trace_header_parser,
         *config.http_bind_address,
-        config.bearer_token,
     )?;
-    let persister = PersisterImpl::new(Arc::clone(&object_store));
+    let persister = Arc::new(PersisterImpl::new(Arc::clone(&object_store)));
     let wal: Option<Arc<WalImpl>> = config
         .wal_directory
         .map(|dir| WalImpl::new(dir).map(Arc::new))
         .transpose()?;
     // TODO: the next segment ID should be loaded from the persister
-    let write_buffer = Arc::new(WriteBufferImpl::new(Arc::new(persister), wal).await?);
-    let query_executor = QueryExecutorImpl::new(
+    let write_buffer = Arc::new(WriteBufferImpl::new(Arc::clone(&persister), wal).await?);
+    let query_executor = Arc::new(QueryExecutorImpl::new(
         write_buffer.catalog(),
         Arc::clone(&write_buffer),
         Arc::clone(&exec),
         Arc::clone(&metrics),
         Arc::new(config.datafusion_config),
         10,
-    );
+    ));
 
     let persister = Arc::new(PersisterImpl::new(Arc::clone(&object_store)));
-    let server = Server::new(
-        common_state,
-        persister,
-        Arc::clone(&write_buffer),
-        Arc::new(query_executor),
-        config.max_http_request_size,
-    );
-    serve(server, frontend_shutdown).await?;
+
+    let builder = ServerBuilder::new(common_state)
+        .max_request_size(config.max_http_request_size)
+        .write_buffer(write_buffer)
+        .query_executor(query_executor)
+        .persister(persister);
+
+    if let Some(token) = config.bearer_token.map(hex::decode).transpose()? {
+        let server = builder
+            .authorizer(Arc::new(AllOrNothingAuthorizer::new(token)))
+            .build();
+        serve(server, frontend_shutdown).await?;
+    } else {
+        let server = builder.build();
+        serve(server, frontend_shutdown).await?;
+    }
```

Comment:
Thanks, I was having a heck of a time with this because by changing the authorizer, it changes the type of `server`, which then, I think, prevents from using a single call to `serve`. I will try these out and see if I can get it to work though, because the separate calls to `serve` are a bit of a smell IMO.

Reply:
Indeed, I still get the same error. For example, with the latter suggestion:

```rust
let server = if let Some(token) = config.bearer_token.map(hex::decode).transpose()? {
    builder
        .authorizer(Arc::new(AllOrNothingAuthorizer::new(token)))
        .build()
} else {
    builder.build()
};

serve(server, frontend_shutdown).await?;
```

I get the following error:

```
error[E0308]: `if` and `else` have incompatible types
   --> influxdb3/src/commands/serve.rs:277:9
    |
272 |        let server = if let Some(token) = config.bearer_token.map(hex::decode).transpose()? {
    |  ___________________-
273 | |/         builder
274 | ||             .authorizer(Arc::new(AllOrNothingAuthorizer::new(token)))
275 | ||             .build()
    | ||____________________- expected because of this
276 | |      } else {
277 | |          builder.build()
    | |          ^^^^^^^^^^^^^^^ expected `AllOrNothingAuthorizer`, found `DefaultAuthorizer`
278 | |      };
    | |______- `if` and `else` have incompatible types
    |
    = note: expected struct `influxdb3_server::Server<_, _, _, AllOrNothingAuthorizer>`
               found struct `influxdb3_server::Server<_, _, _, DefaultAuthorizer>`
```
It is the same issue for the functional style. This is the drawback to the approach I took here. I am apprehensive to resort to dynamic dispatch, but that may be necessary.

One alternative I considered was to just have one `Authorizer`, i.e., the `AllOrNothingAuthorizer`, and make its token optional, e.g.,
```rust
struct AllOrNothingAuthorizer {
    token: Option<Vec<u8>>,
}
```
Then we can construct it once with an optional token, and have a single call to `serve`. Not sure I like that any more, but I'm open to suggestions.

Reply:
I do think avoiding `dyn` is a fine default, but I've also seen it be not that important in some larger scale systems. I think we should do the dynamic dispatch for maintainability and if later we find that it's actually causing perf issues we can fix it, but I think we'll have more bottlenecks when it comes to querying and write perf.

Reply:
I know that Datafusion uses it a fair bit, and it seems to be doing okay 🙂